### ISSUE TITLE
Memory leak in VR_Init

### ISSUE THREAD

**Issue by Kegetys** [ValveSoftware/openvr#195]
VR_Init seems to leak memory every time it is called.

To repro execute the following:

```
while(true)
{
  EVRInitError err = VRInitError_None;
  vr::IVRSystem *pVRSystem = vr::VR_Init(&err, vr::VRApplication_Scene);
  vr::VR_Shutdown();
}
```

And see the process memory use slowly going up.

This is especially a problem if you call VR_Init with VRApplication_Background in a loop to wait for SteamVR to be running, resulting in ever increasing memory use.

**Comment by cryptofbc** [comment]
Have this problem been solved? I have also found this memory leak bug. Is there any method to solve this bug?

**Comment by isometriq** [comment]
I have 5 worker threads that remain alive after per `::init()` call, but only if frames have been submitted to compositor `Compositor()->sumit()`.
Calls to `::shutdown()` does not clean them up.
For each worker thread group of 5, one is priority above normal and the others are idle. The are located on `nvwgf2um.dll` ..which seems to be nvidia geforce.. or cuda stuff?
Anyway ...its leaking!!! help!

**Comment by mann1x** [comment]
Is there anyone that could work on this issue?
Due to this leak is impossible to create any kind of utility or monitoring software using OpenVR.

**Comment by dashingstag** [comment]
Yes I also have the same problem. My workaround is to create a utility tool that calls init only once and have that utility distribute information to the other utilites.

**Comment by mann1x** [comment]
@dashingstag thanks for the info. can you elaborate a bit more?
Do you mean you spawn another process that just calls init once and if it's not successful dies?

**Comment by bddckr** [comment]
SteamVR beta 1.9.7:
>Fixed some memory leaks when VR_Init is called repeatedly in long-running applications.
>https://steamcommunity.com/games/250820/announcements/detail/1714113379461314416

This sadly does _not_ solve the issue. The repo snipped in the first post still does create ~100kB per call (while SteamVR is off at least).

**Comment by mann1x** [comment]
It was much easier to just give up.

I've built a custom driver in a fraction of the time that I've spent on the leak issue, mixing a sample driver with some parts of the driver_leap.

Now on my utility I just have to watch a win32 event handle.
No worries that SteamVR will kill my App cause it missed the ack event or it gets stuck, no leaks.
I can also handle commands via event, like the Recenter for seated position, and notifications on the HMD.

When I'm done I'll publish it together with the C# code for the calls.
Hopefully no one else will have to go through this purgatory again.

**Comment by RotemShi** [comment]
@mann1x did you ever publish the code mentioned above?

**Comment by mann1x** [comment]
@RotemShi whops I didn't realized so much time went by...
Still have to finish a lot of stuff in BSManager before publishing it on Github, last version is broken.

Anyway you can download the latest BSManager version here and check how to use it with dnSpy:

**https://drive.google.com/file/d/10U0ParigpbWuYFl7xaYR7SMjrYi-bpTd/view?usp=sharing**

Just published the repo with the driver:

https://github.com/mann1x/OpenVRmodule

**Comment by RotemShi** [comment]
@mann1x thanks!

**Comment by vilhalmer** [comment]
Still seeing this when SteamVR is waiting for a wireless HMD to connect and thus refuses to initialize.
